core/src/main/java/org/elasticsearch/action/search/RemoteClusterService.java

```diff
@@ -0,0 +1,395 @@
+            }
+            remoteClusters.putAll(this.remoteClusters);
+            for (Map.Entry<String, List<DiscoveryNode>> entry : seeds.entrySet()) {
+                RemoteClusterConnection remote = this.remoteClusters.get(entry.getKey());
+                if (entry.getValue().isEmpty()) { // with no seed nodes we just remove the connection
+                    try {
+                        IOUtils.close(remote);
+                    } catch (IOException e) {
+                        logger.warn("failed to close remote cluster connections for cluster: " + entry.getKey(), e);
+                    }
+                    remoteClusters.remove(entry.getKey());
+                    continue;
+                }
+
+                if (remote == null) { // this is a new cluster we have to add a new representation
+                    remote = new RemoteClusterConnection(settings, entry.getKey(), entry.getValue(), transportService, numRemoteConnections,
+                        nodePredicate);
+                    remoteClusters.put(entry.getKey(), remote);
+                }
+
+                // now update the seed nodes no matter if it's new or already existing
+                RemoteClusterConnection finalRemote = remote;
+                remote.updateSeedNodes(entry.getValue(), ActionListener.wrap(
+                    response -> {
+                        if (countDown.countDown()) {
+                            connectionListener.onResponse(response);
+                        }
+                    },
+                    exception -> {
+                        if (countDown.fastForward()) {
+                            connectionListener.onFailure(exception);
+                        }
+                        if (finalRemote.isClosed() == false) {
+                            logger.warn("failed to update seed list for cluster: " + entry.getKey(), exception);
+                        }
+                    }));
+            }
+        }
+        this.remoteClusters = Collections.unmodifiableMap(remoteClusters);
+    }
+
+    /**
+     * Returns <code>true</code> if at least one remote cluster is configured
+     */
+    boolean isCrossClusterSearchEnabled() {
+        return remoteClusters.isEmpty() == false;
+    }
+
+    /**
+     * Filters out indices that refer to a remote cluster and adds them to the given per cluster indices map.
+     *
+     * @param perClusterIndices a map to fill with remote cluster indices from the given request indices
+     * @param requestIndices the indices in the search request to filter
+     * @param indexExists a predicate that can test if a certain index or alias exists
+     *
+     * @return all indices in the requestIndices array that are not remote cluster indices
+     */
+    public String[] filterIndices(Map<String, List<String>> perClusterIndices, String[] requestIndices, Predicate<String> indexExists) {
```

Comment:
I know we don't like tuples, but I'm also not a fan of methods that mutate input parameters. What do you think of something like:

```diff
diff --git a/core/src/main/java/org/elasticsearch/action/search/RemoteClusterService.java b/core/src/main/java/org/elasticsearch/action/search/RemoteClusterService.java
index 0600d0999e..b98c0868b2 100644
--- a/core/src/main/java/org/elasticsearch/action/search/RemoteClusterService.java
+++ b/core/src/main/java/org/elasticsearch/action/search/RemoteClusterService.java
@@ -57,6 +57,8 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
@@ -175,14 +177,14 @@ public final class RemoteClusterService extends AbstractComponent implements Clo
     /**
      * Filters out indices that refer to a remote cluster and adds them to the given per cluster indices map.
      *
-     * @param perClusterIndices a map to fill with remote cluster indices from the given request indices
+     * @param onRemote       invoked when a request index matches on a remote cluster; the first parameter is the remote cluster alias, the
+     *                       second parameter is the index
+     * @param onLocal        invoked when a request index maches on the local cluster; the parameter is the index
      * @param requestIndices the indices in the search request to filter
-     * @param indexExists a predicate that can test if a certain index or alias exists
-     *
+     * @param indexExists    a predicate that can test if a certain index or alias exists
      * @return all indices in the requestIndices array that are not remote cluster indices
      */
-    public String[] filterIndices(Map<String, List<String>> perClusterIndices, String[] requestIndices, Predicate<String> indexExists) {
-        List<String> localIndicesList = new ArrayList<>();
+    public void filterIndices(BiConsumer<String, String> onRemote, Consumer<String> onLocal, String[] requestIndices, Predicate<String> indexExists) {
         for (String index : requestIndices) {
             int i = index.indexOf(REMOTE_CLUSTER_INDEX_SEPARATOR);
             if (i >= 0) {
@@ -196,21 +198,15 @@ public final class RemoteClusterService extends AbstractComponent implements Clo
                             + remoteCluster + " can't filter indices");
                     }
                     String remoteIndex = index.substring(i + 1);
-                    List<String> indices = perClusterIndices.get(remoteCluster);
-                    if (indices == null) {
-                        indices = new ArrayList<>();
-                        perClusterIndices.put(remoteCluster, indices);
-                    }
-                    indices.add(remoteIndex);
+                    onRemote.accept(remoteCluster, remoteIndex);
                 } else {
-                    localIndicesList.add(index);
+                    onLocal.accept(index);
                 }
             } else {
-                localIndicesList.add(index);
+                onLocal.accept(index);
             }
         }
-        return localIndicesList.toArray(new String[localIndicesList.size()]);
-}
+    }
 
     /**
      * Returns <code>true</code> iff the given cluster is configured as a remote cluster. Otherwise <code>false</code>
diff --git a/core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java b/core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java
index d744148464..24b11b716d 100644
--- a/core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java
+++ b/core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java
@@ -125,8 +125,13 @@ public class TransportSearchAction extends HandledTransportAction<SearchRequest,
         final ClusterState clusterState = clusterService.state();
         if (remoteClusterService.isCrossClusterSearchEnabled()) {
             remoteIndicesByCluster = new HashMap<>();
-            localIndices = remoteClusterService.filterIndices(remoteIndicesByCluster, searchRequest.indices(),
+            final List<String> localIndicesList = new ArrayList<>();
+            remoteClusterService.filterIndices(
+                (r, i) -> remoteIndicesByCluster.computeIfAbsent(r, k -> new ArrayList<>()).add(i),
+                localIndicesList::add,
+                searchRequest.indices(),
                 idx -> indexNameExpressionResolver.hasIndexOrAlias(idx, clusterState));
+            localIndices = localIndicesList.toArray(new String[localIndicesList.size()]);
         } else {
             remoteIndicesByCluster = Collections.emptyMap();
             localIndices = searchRequest.indices();
diff --git a/core/src/test/java/org/elasticsearch/action/search/RemoteClusterServiceTests.java b/core/src/test/java/org/elasticsearch/action/search/RemoteClusterServiceTests.java
index 6c3752d710..09796a724f 100644
--- a/core/src/test/java/org/elasticsearch/action/search/RemoteClusterServiceTests.java
+++ b/core/src/test/java/org/elasticsearch/action/search/RemoteClusterServiceTests.java
@@ -60,6 +60,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
 import java.util.function.Function;
 
 public class RemoteClusterServiceTests extends ESTestCase {
@@ -143,15 +144,20 @@ public class RemoteClusterServiceTests extends ESTestCase {
                     assertTrue(service.isRemoteClusterRegistered("cluster_2"));
                     assertFalse(service.isRemoteClusterRegistered("foo"));
                     Map<String, List<String>> perClusterIndices = new HashMap<>();
-                    String[] localIndices = service.filterIndices(perClusterIndices, new String[]{"foo:bar", "cluster_1:bar",
+                    final List<String> localIndices = new ArrayList<>();
+                    final BiConsumer<String, String> onRemote =
+                        (r, i) -> perClusterIndices.computeIfAbsent(r, k -> new ArrayList<>()).add(i);
+                    service.filterIndices(onRemote, localIndices::add, new String[]{"foo:bar", "cluster_1:bar",
                         "cluster_2:foo:bar", "cluster_1:test", "cluster_2:foo*", "foo"}, i -> false);
-                    assertArrayEquals(new String[]{"foo:bar", "foo"}, localIndices);
+                    assertEquals(2, localIndices.size());
+                    assertEquals("foo:bar", localIndices.get(0));
+                    assertEquals("foo", localIndices.get(1));
                     assertEquals(2, perClusterIndices.size());
                     assertEquals(Arrays.asList("bar", "test"), perClusterIndices.get("cluster_1"));
                     assertEquals(Arrays.asList("foo:bar", "foo*"), perClusterIndices.get("cluster_2"));
 
                     expectThrows(IllegalArgumentException.class, () ->
-                    service.filterIndices(perClusterIndices, new String[]{"foo:bar", "cluster_1:bar",
+                    service.filterIndices(onRemote, localIndices::add, new String[]{"foo:bar", "cluster_1:bar",
                         "cluster_2:foo:bar", "cluster_1:test", "cluster_2:foo*", "foo"}, i -> "cluster_1:bar".equals(i)));
                 }
             }
```